src/doc/trpl/dining-philosophers.md

```diff
@@ -0,0 +1,942 @@
+as an argument and executes that closure in a new thread. This closure needs
+an extra annotation, `move`, to indicate that the closure is going to take
+ownership of the values it’s capturing. Primarily, the `p` variable of the
+`map` function.
+
+Inside the thread, all we do is call `eat()` on `p`.
+
+```rust,ignore
+}).collect();
+```
+
+Finally, we take the result of all those `map` calls and collect them up.
+`collect()` will make them into a collection of some kind, which is why we
+needed to annotate the return type: we want a `Vec<T>`. The elements are the
+return values of the `thread::spawn` calls, which are handles to those threads.
+Whew!
+
+
+```rust,ignore
+for h in handles {
+    h.join().ok().expect("Couldn't join a thread.");
+}
+```
+
+At the end of `main()`, we loop through the handles and call `join()` on them,
+which blocks execution until the thread has completed execution. This ensures
+that the threads complete their work before the program exits.
+
+If you run this program, you’ll see that the philosophers eat out of order!
+We have mult-threading!
+
+```text
+Gilles Deleuze is eating.
+Gilles Deleuze is done eating.
+Friedrich Nietzsche is eating.
+Friedrich Nietzsche is done eating.
+Michel Foucault is eating.
+Baruch Spinoza is eating.
+Baruch Spinoza is done eating.
+Karl Marx is eating.
+Karl Marx is done eating.
+Michel Foucault is done eating.
+```
+
+But what about the forks? We haven’t modeled them at all yet.
+
+To do that, let’s make a new `struct`:
+
+```rust
+use std::sync::Mutex;
+
+struct Table {
+    forks: [Mutex<bool>; 5],
+}
+```
+
+This `Table` has an array with five elements. These elements are all going
+to be `Mutex`es. A mutex is a way to control concurrency: only one thread
+can access the contents at once. This is exactly the property we need
+with our forks. Let’s modify the program to use the `Table`:
+
+```rust
+use std::thread;
+use std::sync::mpsc;
+use std::sync::mpsc::Sender;
+use std::sync::{Mutex, Arc};
+
+struct Philosopher {
+    name: String,
+    done: Sender<bool>,
+    left: usize,
+    right: usize,
+}
+
+impl Philosopher {
+    fn new(name: &str, done: Sender<bool>, left: usize, right: usize) -> Philosopher {
+        Philosopher {
+            name: name.to_string(),
+            done: done,
+            left: left,
+            right: right,
+        }
+    }
+
+    fn done(&self) {
+        println!("{} is done eating.", self.name);
+
+        self.done.send(true).ok().expect("Couldn't finish eating");
+    }
+
+    fn eat(&self, table: &Table) {
+        let _left = table.forks[self.left].lock()
+            .ok().expect("Couldn't aquire left mutex");
+        let _right = table.forks[self.right].lock()
+            .ok().expect("Couldn't aquire right mutex");
+
+        println!("{} is eating.", self.name);
+
+        thread::sleep_ms(1000);
+
+        self.done();
+    }
+}
+
+struct Table {
+    forks: Vec<Mutex<bool>>,
```

Comment:
Also, it looks like the usage here means that the `T` in `Mutex<T>` can just be `()` I believe?